### ISSUE TITLE
Bug: BackstopPool does not emit InsuranceFeeSet

### ISSUE THREAD

**Issue by TorstenStueber** [pendulum-chain/wasm-deploy#46]
The [current Nabla deployment on Foucoco](https://www.notion.so/satoshipay/24-02-17-Nabla-Launch-b637052622c24e3baf68964af1ecd7fb?pvs=4#c8d9c0933b4a485384ce10e885a3149e) has the backstop pool at this address: `6hZJZnisXo5LeFwV3aE4piaJCCPeRTTf9iiASNFTwYc1jVDP`.

This deployment uses [this version](https://github.com/0xamberhq/contracts/blob/pendulum-launch/contracts/src/core/BackstopPool.sol) of the backstop pool.

It is obvious that the method `setInsuranceFee` should emit the `InsuranceFeeSet` event whenever it runs successfully.

However, this is not the case. When, e.g., calling `setInsuranceFee` as follows:
* from the owner account (`6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3`)
* for the first swap pool (`6hJ5MDmQohfjxmyXbxNvCqaiousw9uqwjKMAH8JGe7GfUMow`)
* for any `insuranceFeeBps` (< 3000)
then the call is successful (which can be verified by reading `getInsuranceFee` afterwards, but no event is emitted.

**Comment by TorstenStueber** [comment]
This problem could be due to a compilation problem with Solang (?)

It's effect is that the indexer cannot correctly track the insurance fee of a swap pool.

**Comment by ebma** [comment]
I don't think I'm able to reproduce this issue. I deployed the contracts on Foucoco using the same deployment account as you did. As you can see in the following logs, the `InsuranceFeeSet` event is emitted when using an `insuranceFeeBps` of `0`. I did not change this parameter in the default configuration.

I'm using the contracts from the `pendulum-launch` branch and deploying from `39-implement-tests-for-nabla-indexer`. I compiled solang v0.3.3 from source. 

```
 🧰 backstop.addSwapPool [ fee: 0.001 AMPE ] success
    🎉 Event backstop.InsuranceFeeSet
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6kgKPJbCQKyDVmyFMMkESkT2JkYfaXFqfr4CqxBL9Wz7idjD"
      - insuranceFeeBps: "0"
    🎉 Event backstop.SwapPoolAdded
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6kgKPJbCQKyDVmyFMMkESkT2JkYfaXFqfr4CqxBL9Wz7idjD"
  🧰 router.registerPool [ fee: 0.0017 AMPE ] success
    🎉 Event xcm0Erc20Wrapper.Approval
      - owner: "6mwyRVmxj3SDoinATY36L7rfnKNaqkJboSMUkEXoa3hb5e2Q"
      - spender: "6jWmBYexLnjiqjtMLT3diRmbGRRd1KjA5uTw8MResQe2fB5w"
      - value: "115,792,089,237,316,195,423,570,985,008,687,907,853,269,984,665,640,564,039,457,584,007,913,129,639,935"
    🎉 Event router.SwapPoolRegistered
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - pool: "6jWmBYexLnjiqjtMLT3diRmbGRRd1KjA5uTw8MResQe2fB5w"
      - asset: "6iD9m1iDPTqEeesBuWckVfS3EWhKkGGSLnKSzLwF1bjeCVYW"
  🧰 backstop.addSwapPool [ fee: 0.001 AMPE ] success
    🎉 Event backstop.InsuranceFeeSet
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6jWmBYexLnjiqjtMLT3diRmbGRRd1KjA5uTw8MResQe2fB5w"
  🧰 router.registerPool [ fee: 0.0017 AMPE ] success
    🎉 Event nativeErc20Wrapper.Approval
      - owner: "6mwyRVmxj3SDoinATY36L7rfnKNaqkJboSMUkEXoa3hb5e2Q"
      - spender: "6kgKPJbCQKyDVmyFMMkESkT2JkYfaXFqfr4CqxBL9Wz7idjD"
      - value: "115,792,089,237,316,195,423,570,985,008,687,907,853,269,984,665,640,564,039,457,584,007,913,129,639,935"
    🎉 Event router.SwapPoolRegistered
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - pool: "6kgKPJbCQKyDVmyFMMkESkT2JkYfaXFqfr4CqxBL9Wz7idjD"
      - asset: "6juXMwQSBybzNAz1i9kvWHnfkvPh5fKi6F11EjbkMXuojaJ8"
  🧰 backstop.addSwapPool [ fee: 0.001 AMPE ] success
    🎉 Event backstop.InsuranceFeeSet
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6kgKPJbCQKyDVmyFMMkESkT2JkYfaXFqfr4CqxBL9Wz7idjD"
      - insuranceFeeBps: "0"
    🎉 Event backstop.SwapPoolAdded
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6kgKPJbCQKyDVmyFMMkESkT2JkYfaXFqfr4CqxBL9Wz7idjD"
  🧰 router.registerPool [ fee: 0.0017 AMPE ] success
    🎉 Event xcm0Erc20Wrapper.Approval
      - owner: "6mwyRVmxj3SDoinATY36L7rfnKNaqkJboSMUkEXoa3hb5e2Q"
      - spender: "6jWmBYexLnjiqjtMLT3diRmbGRRd1KjA5uTw8MResQe2fB5w"
  🧰 backstop.addSwapPool [ fee: 0.001 AMPE ] success
    🎉 Event backstop.InsuranceFeeSet
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6kgKPJbCQKyDVmyFMMkESkT2JkYfaXFqfr4CqxBL9Wz7idjD"
      - insuranceFeeBps: "0"
    🎉 Event backstop.SwapPoolAdded
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6kgKPJbCQKyDVmyFMMkESkT2JkYfaXFqfr4CqxBL9Wz7idjD"
  🧰 router.registerPool [ fee: 0.0017 AMPE ] success
    🎉 Event xcm0Erc20Wrapper.Approval
      - owner: "6mwyRVmxj3SDoinATY36L7rfnKNaqkJboSMUkEXoa3hb5e2Q"
      - spender: "6jWmBYexLnjiqjtMLT3diRmbGRRd1KjA5uTw8MResQe2fB5w"
      - value: "115,792,089,237,316,195,423,570,985,008,687,907,853,269,984,665,640,564,039,457,584,007,913,129,639,935"
    🎉 Event router.SwapPoolRegistered
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - pool: "6jWmBYexLnjiqjtMLT3diRmbGRRd1KjA5uTw8MResQe2fB5w"
      - asset: "6iD9m1iDPTqEeesBuWckVfS3EWhKkGGSLnKSzLwF1bjeCVYW"
  🧰 backstop.addSwapPool [ fee: 0.001 AMPE ] success
    🎉 Event backstop.InsuranceFeeSet
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6kgKPJbCQKyDVmyFMMkESkT2JkYfaXFqfr4CqxBL9Wz7idjD"
      - insuranceFeeBps: "0"
    🎉 Event backstop.SwapPoolAdded
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6kgKPJbCQKyDVmyFMMkESkT2JkYfaXFqfr4CqxBL9Wz7idjD"
  🧰 router.registerPool [ fee: 0.0017 AMPE ] success
    🎉 Event xcm0Erc20Wrapper.Approval
      - owner: "6mwyRVmxj3SDoinATY36L7rfnKNaqkJboSMUkEXoa3hb5e2Q"
      - spender: "6jWmBYexLnjiqjtMLT3diRmbGRRd1KjA5uTw8MResQe2fB5w"
      - value: "115,792,089,237,316,195,423,570,985,008,687,907,853,269,984,665,640,564,039,457,584,007,913,129,639,935"
    🎉 Event router.SwapPoolRegistered
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - pool: "6jWmBYexLnjiqjtMLT3diRmbGRRd1KjA5uTw8MResQe2fB5w"
      - asset: "6iD9m1iDPTqEeesBuWckVfS3EWhKkGGSLnKSzLwF1bjeCVYW"
  🧰 backstop.addSwapPool [ fee: 0.001 AMPE ] success
    🎉 Event backstop.InsuranceFeeSet
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6jWmBYexLnjiqjtMLT3diRmbGRRd1KjA5uTw8MResQe2fB5w"
      - insuranceFeeBps: "0"
    🎉 Event backstop.SwapPoolAdded
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6jWmBYexLnjiqjtMLT3diRmbGRRd1KjA5uTw8MResQe2fB5w"
  🧰 router.registerPool [ fee: 0.0017 AMPE ] success
    🎉 Event xcm1Erc20Wrapper.Approval
      - owner: "6mwyRVmxj3SDoinATY36L7rfnKNaqkJboSMUkEXoa3hb5e2Q"
      - spender: "6nPPULtZ73qYz8se3nvYSvLHmfadLffFKAawooYqbryUH4Uz"
      - value: "115,792,089,237,316,195,423,570,985,008,687,907,853,269,984,665,640,564,039,457,584,007,913,129,639,935"
    🎉 Event router.SwapPoolRegistered
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - pool: "6nPPULtZ73qYz8se3nvYSvLHmfadLffFKAawooYqbryUH4Uz"
      - asset: "6mQekcLz1xEDW2YTcMQWPoPryhEi3ZxANXokfsz1JzXei7aR"
  🧰 backstop.addSwapPool [ fee: 0.001 AMPE ] success
    🎉 Event backstop.InsuranceFeeSet
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6nPPULtZ73qYz8se3nvYSvLHmfadLffFKAawooYqbryUH4Uz"
      - insuranceFeeBps: "0"
    🎉 Event backstop.SwapPoolAdded
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6nPPULtZ73qYz8se3nvYSvLHmfadLffFKAawooYqbryUH4Uz"
```

**Comment by TorstenStueber** [comment]
This makes a lot of sense but makes it just more confusing. `InsuranceFeeSet` is emitted inside [`_setInsuranceFee`](https://github.com/0xamberhq/contracts/blob/0a80ae5cb2d015787206b315703430920ca1728a/contracts/src/core/BackstopPool.sol#L127C14-L127C30). 

This function is called both in [`addSwapPool`](https://github.com/0xamberhq/contracts/blob/0a80ae5cb2d015787206b315703430920ca1728a/contracts/src/core/BackstopPool.sol#L92) as well as [`setInsuranceFee`](https://github.com/0xamberhq/contracts/blob/0a80ae5cb2d015787206b315703430920ca1728a/contracts/src/core/BackstopPool.sol#L117).

The deployment code shows that `InsuranceFeeSet` is correctly emitted when `addSwapPool` is called. However, this issue here particularly refers to the other case, namely that `InsuranceFeeSet` does not seem to be emitted when `setInsuranceFee` is called.

**Comment by TorstenStueber** [comment]
I just tried it myself using the wasm-deploy script and added a call to `backstop.setInsuranceFee`. And indeed this works as required:

```
  🧰 backstop.setInsuranceFee [ fee: 0.0006 AMPE ] success
    🎉 Event backstop.InsuranceFeeSet
      - sender: "6iFKMtX29zYoHRgDkPTXKuRsHRbJ3Gnaxyc4dRrZTANHvZi3"
      - swapPool: "6jyMMKi1UA3hszW54NC79vcyJLNwbSktTF4XotoDeveeAAej"
      - insuranceFeeBps: "1,234"
```

**Comment by TorstenStueber** [comment]
Oh, I just realized that the contract at `6h4rVhtWSJcZdxKspLnM7LpjgGdmTUTPSnLqpoYia2aR6vqc` does not have the same code hash, according to the pallet-contract storage it is the code hash `0x026ff1b46ca6ae1d60aed9cea110e48ee635da3087675e463bc717a9c23ffbe1`.

That explains why the contract is not working as expected. I will check what version of the backstop source code compiles to this code.

That means I did not deploy the latest version of the backstop pool contract. Either I did something wrong or there is some problem in wasm-deploy that it did not properly pull and compile the latest version. Hard to track down what actually happened.

We can probably close this ticket now.

**Comment by TorstenStueber** [comment]
I left a bunch of messages here but I deleted them again. After doing more tests I finally realized that I just used linked to an old version of the backstop pool when adding the backstop pool to polkadot.js. I linked to the backstop pool at address `6h4rVhtWSJcZdxKspLnM7LpjgGdmTUTPSnLqpoYia2aR6vqc` instead of `6hZJZnisXo5LeFwV3aE4piaJCCPeRTTf9iiASNFTwYc1jVDP`. The backstop pool at the former address uses [this version of the code](https://github.com/0xamberhq/contracts/blob/3e96787e161fb2657c90f18034691749647b691f/contracts/src/core/BackstopPool.sol#L122-L131) where `setInsuranceFee` does not emit an event.

I now tested with the backstop pool at the correct address and it works as expected and the indexer picks up the change as expected.

I will close this issue now.
